<h2>SANTOS, STEPHEN CHRISTIAN</h2>
<h2>BSIT-3A</h2>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [2]:
accidents = pd.read_csv('uk_road_accident.csv')

In [3]:
accidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accidents['Accident Date'] = accidents['Accident Date'].str.strip()

accidents['Accident Date'] = accidents['Accident Date'].astype('str')

accidents['Accident Date'] = accidents['Accident Date'].str.replace('/', '-')

accidents['Accident Date'] = pd.to_datetime(accidents['Accident Date'], dayfirst=True, errors='coerce')

In [5]:
accidents['Latitude'] = accidents['Latitude'].fillna(accidents['Latitude'].mode()[0])
accidents['Longitude'] = accidents['Longitude'].fillna(accidents['Longitude'].mode()[0])
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].fillna('unaccounted')
accidents['Road_Type'] = accidents['Road_Type'].fillna('unaccounted')
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].fillna('unaccounted')
accidents['Weather_Conditions'] = accidents['Weather_Conditions'].fillna(accidents['Weather_Conditions'].mode()[0])

In [6]:
accidents['Year'] = accidents['Accident Date'].dt.year
accidents['Month'] = accidents['Accident Date'].dt.month
accidents['Day'] = accidents['Accident Date'].dt.day
accidents['DayOfWeek'] = accidents['Accident Date'].dt.dayofweek
accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

<h2>1. Which district area has the highest number of accidents?</h2>

In [7]:
accidents['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h2>INSIGHTS #1</h2>
<p>Birmingham has the highest number of accidents with 13,491.</p>

<h2>2. What is the average number of vehicles involved in accidents for each Road Type?</h2>

In [8]:
accidents.groupby('Road_Type')['Number_of_Vehicles'].mean()

Road_Type
Dual carriageway      2.009766
One way street        1.585441
Roundabout            1.901323
Single carriageway    1.795911
Slip road             1.917483
unaccounted           1.674115
Name: Number_of_Vehicles, dtype: float64

<h2>INSIGHTS #2</h2>
<p>The average number of vehicles involved in accidents is highest on dual carriageways with about 2 vehicles, followed by slip roads with about 1.92, roundabouts with about 1.90, single carriageways with about 1.80, unaccounted roads with about 1.67, and the lowest is on one-way streets with about 1.59 vehicles.</p>

<h2>3. How many accidents happened in each Light Conditions category?</h2>

In [9]:
accidents['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h2>INSIGHTS #3</h2>
<p>There were 484,880 accidents in daylight, 129,335 in darkness with street lights on, 37,437 in darkness with no lighting, 6,484 in darkness with unknown lighting, and 2,543 in darkness with street lights off.</p>

<h2>4. What is the total number of casualties for each Vehicle Type?</h2>

In [10]:
accidents.groupby('Vehicle_Type')['Number_of_Casualties'].sum()

Vehicle_Type
Agricultural vehicle                       2613
Bus or coach (17 or more pass seats)      34915
Car                                      676692
Data missing or out of range                  9
Goods 7.5 tonnes mgw and over             23397
Goods over 3.5t. and under 7.5t            8308
Minibus (8 - 16 passenger seats)           2659
Motorcycle 125cc and under                20348
Motorcycle 50cc and under                 10167
Motorcycle over 125cc and up to 500cc     10286
Motorcycle over 500cc                     34879
Other vehicle                              7554
Pedal cycle                                 270
Ridden horse                                  5
Taxi/Private hire car                     18195
Van / Goods 3.5 tonnes mgw or under       46271
Name: Number_of_Casualties, dtype: int64

<h2>INSIGHTS #4</h2>
<p>Cars had the highest number with 676,692 casualties.
Buses or coaches had 34,915, while vans or goods vehicles under 3.5 tonnes had 46,271.
Motorcycles over 500cc had 34,879, followed by goods vehicles over 7.5 tonnes with 23,397.
Motorcycles 125cc and under had 20,348, and taxi/private hire cars had 18,195.
Motorcycles over 125cc and up to 500cc had 10,286, and motorcycles 50cc and under had 10,167.
Goods vehicles between 3.5 and 7.5 tonnes had 8,308, and other vehicles had 7,554.
Minibuses had 2,659, and agricultural vehicles had 2,613.
Pedal cycles had 270, ridden horses had 5, and data missing or out of range had 9 casualties.</p>

<h2>5. How many accidents occurred on each Road Type?</h2>

In [11]:
accidents['Road_Type'].value_counts()

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
unaccounted             4520
Name: count, dtype: int64

<h2>INSIGHTS #5</h2>
<p>There were 492,143 accidents on single carriageways, 99,424 on dual carriageways, 43,992 on roundabouts, 13,559 on one-way streets, 7,041 on slip roads, and 4,520 on unaccounted road types.</p>

<h2>6. What is the average number of casualties per accident in Urban vs Rural Areas?</h2>

In [12]:
accidents.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287815
unaccounted    1.066667
Name: Number_of_Casualties, dtype: float64

<h2>INSIGHTS #6</h2>
<p>The average number of casualties per accident is higher in rural areas with about 1.48 casualties, compared to urban areas, which have about 1.29 casualties per accident.</p>

<h2>7. How many accidents happened with each Weather Condition?</h2>

In [13]:
accidents['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2>INSIGHTS #7</h2>
<p>There were 535,013 accidents in fine weather with no high winds, 79,696 in rain without high winds, 17,150 in other weather conditions, 9,615 in rain with high winds, 8,554 in fine weather with high winds, 6,238 in snow without high winds, 3,528 in fog or mist, and 885 in snow with high winds.</p>

<h2>INSIGHTS #8</h2>
<p>Most accidents happen during clear weather, whether it’s calm or windy, over 80% of all accidents occurred in fine weather conditions (combining both fine with no high winds and fine with high winds). This suggests that bad weather isn’t the main cause of most accidents, many crashes still happen when the weather is good, likely due to other factors like driver behavior, traffic volume, or road design.</p>

<h2>8. How many accidents involved more than 2 vehicles?</h2>

In [14]:
accidents[accidents['Number_of_Vehicles'] > 2].value_counts().sum()

np.int64(67897)

<h2>INSIGHTS #9</h2>
<p>There were 67,897 accidents that involved more than 2 vehicles.</p>

<h2>9. What is the distribution of Accident Severity in the dataset?</h2>

In [15]:
accidents['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h2>INSIGHTS #10</h2>
<p>Slight has 563,801 cases. There are fewer serious accidents, totaling 88,217, and the least common are fatal accidents, with 8,661 reported.</p>

<h2>10. What is the total number of casualties for each Light Conditions?</h2>

In [16]:
accidents.groupby('Light_Conditions')['Number_of_Casualties'].sum()

Light_Conditions
Darkness - lighting unknown      8471
Darkness - lights lit          177108
Darkness - lights unlit          3385
Darkness - no lighting          57715
Daylight                       649889
Name: Number_of_Casualties, dtype: int64

<h2>INSIGHTS #11</h2>
<p>In daylight, there were 649,889 casualties. In darkness with lights on, there were 177,108 casualties. In darkness with no lighting, there were 57,715 casualties. In darkness with lighting unknown, there were 8,471 casualties. In darkness with lights off, there were 3,385 casualties.</p>

<h2>INSIGHTS #12</h2>
<p>Even though most accidents happen during daylight, the number of casualties in darkness (whether lights are on or off) is still very high, showing that accidents at night tend to be more serious or cause more injuries compared to those in the daytime.</p>

<h2>11. How many accidents occurred on each day of the week (using DayOfWeek)?</h2>

In [17]:
accidents.groupby('DayOfWeek').size()

DayOfWeek
0     72680
1     94550
2     99558
3     99511
4     97900
5    107178
6     89302
dtype: int64

<h2>INSIGHTS #13</h2>
<p>The highest number of accidents happened on day 5 with 107,178 accidents, while the lowest was on day 0 with 72,680 accidents. The other days had numbers between these, with days 2 and 3 both having just under 100,000 accidents, and day 6 having around 89,000 accidents.</p>

<h2>12. What is the average number of casualties per accident on each day of the week?</h2>

In [18]:
accidents.groupby('DayOfWeek')['Number_of_Casualties'].mean()

DayOfWeek
0    1.465685
1    1.331793
2    1.317925
3    1.321160
4    1.323238
5    1.350977
6    1.423272
Name: Number_of_Casualties, dtype: float64

<h2>INSIGHTS #14</h2>
<p>The average number of casualties per accident is highest on day 0 with about 1.47 casualties, and lowest on day 2 with about 1.32 casualties. The other days have averages close to these, ranging between about 1.32 and 1.35 casualties per accident.</p>

<h2>13. What is the Number of accidents by Vehicle Type and Accident Severity?</h2>

In [25]:
accidents.groupby(['Vehicle_Type', 'Accident_Severity']).size().unstack()

Accident_Severity                       Fatal  Serious    Slight
Vehicle_Type                                                    
Agricultural vehicle                     21.0    282.0    1644.0
Bus or coach (17 or more pass seats)    325.0   3373.0   22180.0
Car                                    6577.0  66461.0  424954.0
Data missing or out of range              NaN      NaN       6.0
Goods 7.5 tonnes mgw and over           216.0   2321.0   14770.0
Goods over 3.5t. and under 7.5t          67.0    857.0    5172.0
Minibus (8 - 16 passenger seats)         29.0    276.0    1671.0
Motorcycle 125cc and under              189.0   2031.0   13049.0
Motorcycle 50cc and under                95.0   1014.0    6494.0
Motorcycle over 125cc and up to 500cc   105.0   1014.0    6537.0
Motorcycle over 500cc                   339.0   3457.0   21861.0
Other vehicle                            70.0    767.0    4800.0
Pedal cycle                               6.0     39.0     152.0
Ridden horse                              NaN      NaN       4.0
Taxi/Private hire car                   155.0   1771.0   11368.0
Van / Goods 3.5 tonnes mgw or under     467.0   4554.0   29139.0

<h2>INSIGHTS #15</h2>
<p>Most accidents with cars and buses caused slight injuries. Cars had the most slight injuries, then serious, and fewer fatal accidents. Motorcycles and goods vehicles also had mostly slight injuries, with some serious and fewer fatal cases. Overall, most accidents caused slight injuries, while serious and fatal accidents were less common.</p>
<h2>INSIGHTS #16</h2>
<p>Most accidents cause slight injuries, vehicles like motorcycles and taxis have a higher chance of serious and fatal accidents compared to cars and buses.</p>
<h2>INSIGHTS #17</h2>
<p>Vehicles carrying passengers, like buses and minibuses, have fewer fatal accidents compared to goods vehicles and motorcycles, even though they are involved in many accidents.</p>

<h2>14. What is the Number of casualties by Aberdeen City and Accident Severity?</h2>

In [26]:
accidents.groupby(['District Area', 'Accident_Severity'])['Number_of_Casualties'].sum()

District Area  Accident_Severity
Aberdeen City  Fatal                  19
               Serious               304
               Slight               1185
Aberdeenshire  Fatal                 120
               Serious               686
                                    ... 
Wyre Forest    Serious               201
               Slight               1093
York           Fatal                  33
               Serious               385
               Slight               2035
Name: Number_of_Casualties, Length: 1266, dtype: int64

<h2>INSIGHTS #18</h2>
<p>In Aberdeen City, there were 19 fatal casualties, 304 serious casualties, and 1,185 slight casualties.</p>

<h2>INSIGHTS #19</h2>
<p>Tt seems that in many areas, slight injuries make up the largest share of casualties, while fatal injuries are the least common. This pattern shows that most accidents tend to cause minor injuries rather than serious or deadly ones across different districts.</p>

<h2>INSIGHTS #20</h2>
<p>Some areas like Aberdeenshire, the number of serious casualties is much higher compared to fatal ones, which might mean that accidents there often result in injuries that need medical attention but are less likely to be deadly.</p>

<h2>15.	What is the Average number of vehicles involved in accidents by Vehicle Type?</h2>

In [21]:
accidents.groupby('Vehicle_Type')['Number_of_Vehicles'].mean()

Vehicle_Type
Agricultural vehicle                     1.805855
Bus or coach (17 or more pass seats)     1.831015
Car                                      1.832134
Data missing or out of range             1.666667
Goods 7.5 tonnes mgw and over            1.830935
Goods over 3.5t. and under 7.5t          1.833497
Minibus (8 - 16 passenger seats)         1.804150
Motorcycle 125cc and under               1.815247
Motorcycle 50cc and under                1.839537
Motorcycle over 125cc and up to 500cc    1.825888
Motorcycle over 500cc                    1.834743
Other vehicle                            1.825971
Pedal cycle                              1.837563
Ridden horse                             1.500000
Taxi/Private hire car                    1.836468
Van / Goods 3.5 tonnes mgw or under      1.824180
Name: Number_of_Vehicles, dtype: float64

<h2>INSIGHTS #21</h2>
<p>The average number of vehicles involved in accidents is quite similar across most vehicle types, usually around 1.8 to 1.83 vehicles. The lowest average is for ridden horses with 1.5 vehicles, while motorcycles and pedal cycles tend to be involved in slightly more multi-vehicle accidents compared to other types.</p>

<h2>INSIGHTS #22</h2>
<p>Most vehicle types tend to be involved in accidents with about two vehicles, which means many accidents happen between just one or two vehicles rather than large multi-vehicle crashes.</p>

<h2>16. Do accidents at darkness to be more severe without lighting?</h2>

In [22]:
accidents.groupby('Light_Conditions')['Accident_Severity'].value_counts()

Light_Conditions             Accident_Severity
Darkness - lighting unknown  Slight                 5622
                             Serious                 794
                             Fatal                    68
Darkness - lights lit        Slight               108345
                             Serious               19130
                             Fatal                  1860
Darkness - lights unlit      Slight                 2138
                             Serious                 360
                             Fatal                    45
Darkness - no lighting       Slight                28651
                             Serious                7174
                             Fatal                  1612
Daylight                     Slight               419045
                             Serious               60759
                             Fatal                  5076
Name: count, dtype: int64

<h2>INSIGHTS #23</h2>
<p>Accidents in darkness without lighting have fewer slight injuries but much more serious and fatal injuries compared to accidents where lights are on, showing that lack of lighting makes accidents more severe.</p>
<h2>INSIGHTS #24</h2>
<p>Based on the data, even though accidents happen less often in darkness with unknown lighting, they still cause a noticeable number of serious and fatal injuries.</p>

<h2>17.	What is the average number of vehicles involved in accidents?</h2>

In [23]:
accidents['Number_of_Vehicles'].mean()

np.float64(1.8312554205597575)

<h2>INSIGHTS #25</h2>
<p>The average number of vehicles involved in each accident is about 1.83.</p>